scipy/optimize/tests/test_nnls.py

```diff
@@ -42,3 +42,53 @@ def test_maxiter(self):
         b = self.rng.uniform(size=5)
         with assert_raises(RuntimeError):
             nnls(a, b, maxiter=1)
+
+
+    def test_nnls_max_iterations(self):
+        # This is from a real test case that failed due to a problem in the scipy._nnls implementation
+        n = np.array([3, 2, 0, 1, 1, 1, 3, 8, 14, 16, 29, 23, 41,
+                      47, 53, 57, 67, 76, 103, 89, 97, 94, 85, 95, 78, 78,
+                      78, 77, 73, 50, 50, 56, 68, 98, 95, 112, 134, 145, 158,
+                      172, 213, 234, 222, 215, 216, 216, 206, 183, 135, 156, 110, 92,
+                      63, 60, 52, 29, 20, 16, 12, 5, 5, 5, 1, 2, 3,
+                      0, 2])
+        k = np.array([0., 0., 0., 0., 0., 0.,
+                      0., 0., 0.,
+                      0., 0., 0.,
+                      0., 0., 0.,
+                      0., 0., 0.,
+                      0., 0., 0.7205812007860187,
+                      0., 1.4411624015720375, 0.7205812007860187,
+                      2.882324803144075, 5.76464960628815, 5.76464960628815,
+                      12.249880413362318, 15.132205216506394, 20.176273622008523,
+                      27.382085629868712, 48.27894045266326, 47.558359251877235,
+                      68.45521407467177, 97.99904330689854, 108.0871801179028,
+                      135.46926574777152, 140.51333415327366, 184.4687874012208,
+                      171.49832578707245, 205.36564222401535, 244.27702706646033,
+                      214.01261663344755, 228.42424064916793, 232.02714665309804,
+                      205.36564222401535, 172.9394881886445, 191.67459940908097,
+                      162.1307701768542, 153.48379576742198, 110.96950492104689,
+                      103.04311171240067, 86.46974409432225, 60.528820866025576,
+                      43.234872047161126, 23.779179625938617, 24.499760826724636,
+                      17.29394881886445, 11.5292992125763, 5.76464960628815,
+                      5.044068405502131, 3.6029060039300935, 0.,
+                      2.882324803144075, 0., 0.,
+                      0.])
+        # The following code sets up a system of equations such that $k_i - p_i * n_i$ is minimized for $p_i$ with
+        # weights $n_i$ and monotonicity constraints on $p_i$. This translates to a system of equations of the form
+        # $k_i - (d_1 + ... + d_i) * n_i$ and non-negativity constraints on the $d_i$. If $n_i$ is zero the system is
+        # modified such that $d_i - d_{i+1}$ is then minimized.
+        N = len(n)
+        A = np.diag(n) @ np.tril(np.ones((N, N)))
+        w = n ** 0.5
+
+        nz = (n == 0).nonzero()[0]
+        A[nz, nz] = 1
+        A[nz, np.minimum(nz + 1, N - 1)] = -1
+        w[nz] = 1
+        k[nz] = 0
+        W = np.diag(w)
+
+        # Small perturbations can already make the infinity loop go away (just uncomment the next line)
+        # k = k + 1e-10 * np.random.normal(size=N)
+        d, _ = nnls(W @ A, W @ k)
```

Comment:
We need to compare this or any other expected outcomes with the actual results to make a test or fail automatically. Could you please add the correctness test here?

Reply:
I added the expected result and made all lines fit into 80 characters. So that should be resolved.

Just one more thing I noticed: when I run this unit test in scipy, it does not fail because of the `RuntimeError("Maximum number of iterations reached.")`, but because of some `RuntimeWarning: invalid value encountered in divide`. This happens because in the line `alpha = (x[alpha_ind] / (x[alpha_ind] - s[alpha_ind])).min()  # C.2` there's a division `0./0.`, which should give a `nan` (which is the case when I run the test in my project) but raises an exception here. I guess that's because the RuntimeWarning is promoted to an error by some `filterwarnings` stuff or pytest setting, but I'm not quite sure about that.

Reply:
That's I think another bug in the line above it, taking also 0 values. Hence if we don't use tol values then 

```python
alpha_ind = (s[P] < 0).nonzero()
```

then also below it, we probably need

```python
P[x <= 0] = False
```

Reply:
I fixed that, but then I generated some more test data and for some test cases it still fails. It now quits the inner loop after having removed some passive variables, but then enters it again after having set them passive again. E.g. in the new unit test I pushed, when I trace the passive variables I get this (just an extract, so this doesn't get too long)
```
<snip>
Outer loop:  passive=[ 0 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64]
Outer loop:  passive=[ 0 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0  1 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0  1  2 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0  1  2  3 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0  1  2  3  4 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0  1  2  3  4  5 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Inner loop:  passive=[22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0  1 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0  1  2 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0  1  2  3 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0  1  2  3  4 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0  1  2  3  4  5 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Inner loop:  passive=[22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
Outer loop:  passive=[ 0 22 25 28 29 30 31 32 33 34 35 36 48 49 55 64 67]
</snip>
```
The indices 0 to 5 are added one by one to the passive set in the outer loop, but somehow as soon as number 5 is added, they become (slightly) negative and all are removed in the inner loop. Then the whole cycle starts anew in the outer loop.

I haven't found a clean way to solve this yet, except currently just setting the tolerance higher.

Reply:
Probably we need to get it negative enough so maybe using `s[P] < -tol` ?

Reply:
I tried that but that didn't fix it either. It still went into some infinite loop.
Would be okay to put those variables, that were set passive and then active again, into some kind of "permanent active" set, such that they just stay there and won't get added (and removed) again? But I don't know whether there are cases where this happens, say two times, and then stabilizes and leads to the termination of the algorithm and the correct solution.
Maybe your solution with `-tol` would work but with another value of `tol`? I'm not sure, but it could be related to the condition number of the sub-matrix of A given by the passive set, but it would take some time to come up with a good theory here, to support that. 
Furthermore, we could run into trouble because at termination some variables could be slightly negative and some algorithms which need `nnls` are fine with that, but others aren't and need complete non-negativity, maybe sacrificing some optimality for it.
Just my two cents...

Reply:
I am just trying to understand why that is the case so probably I am translating the algorithm incorrectly because according to the authors the set should shrink monotonously. I will have some time this weekend to sit down and read the paper.